### ISSUE TITLE
Long URLs/query params fail with no explanation

### ISSUE THREAD

**Issue by ThomWright** [linkerd/linkerd2#4322]
### What is the issue?

Requests to URLs with really long query params seem to fail in a difficult-to-debug way.

Symptoms (some of which _might_ be unrelated) for requests from `C` to `S`:

- client `C` reporting 502s
- linkerd-proxy on the client reporting errors: `connection closed before message completed` (see below)
- grafana dashboard the for the svc `S` showing spikes of 50sec latency
- grafana dashboard for the deployment `S` showing normal latency

### How can it be reproduced?

Send some requests with shit tons of query params. We found a request with 8677 characters which exhibited the failure.

### Logs, error output, etc

```
ERR! [  1236.631544s] linkerd2_app_core::errors unexpected error: connection closed before message completed
WARN [  1237.800060s] hyper::proto::h1::role response with HTTP2 version coerced to HTTP/1.1    
WARN [  1308.540734s] hyper::proto::h1::role response with HTTP2 version coerced to HTTP/1.1    
ERR! [  1359.964981s] linkerd2_app_core::errors unexpected error: connection closed before message completed
WARN [  1361.192850s] hyper::proto::h1::role response with HTTP2 version coerced to HTTP/1.1
```

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API
kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version
linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist
linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API
linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ no invalid service profiles
linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 2.6.1 but the latest stable version is 2.7.1
    see https://linkerd.io/checks/#l5d-version-cli for hints
control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 2.6.1 but the latest stable version is 2.7.1
    see https://linkerd.io/checks/#l5d-version-control for hints
√ control plane and cli versions match
Status check results are √
```

### Environment

- Kubernetes Version: 1.15.11-gke.3
- Cluster Environment: GKE
- Host OS: Container-Optimized OS from Google
- Linkerd version: 2.6.1

### Possible solution

- Improved error messaging about URLs which are too long?
- Immediate request termination for URLs which are too long?

### Additional context

I'm still unsure exactly where the request fails. I don't know if it's in Linkerd or somewhere in between! So it's possible it's not actually a Linkerd problem.

Original slack thread: https://linkerd.slack.com/archives/C89RTCWJF/p1588343181485300

**Comment by grampelberg** [comment]
I wonder what the RFC says the max length of an URL can be.

**Comment by olix0r** [comment]
@grampelberg IIRC the RFC doesn't expose a hard limit. Implementations have to choose max buffer sizes and error accordingly. I believe we can modify our buffer size, though...

**Comment by grampelberg** [comment]
@olix0r maybe just exposing an error when we hit the limit would be enough?

**Comment by olix0r** [comment]
@grampelberg It sounds like we should be returning a 414 in this scenario...

**Comment by ThomWright** [comment]
A 414 would be great.

**Comment by zaharidichev** [comment]
@grampelberg  @olix0r  So what is happening here is interesting. This situation occurs when we are going over  H2. So if one of the pods is not meshed we are not going to be upgrading the protocol and things will work out fine. However when we upgrade to H2, we do get a `UserError::HeaderTooBig` because our `path` pseudo header is too big. This happens in H2 right [here](https://github.com/hyperium/h2/blob/master/src/frame/headers.rs#L962). We cannot properly introspect this error and we simply derive 502 from it. Will fix it

**Comment by zaharidichev** [comment]
Filed an issue in H2 that concerns this problem: https://github.com/hyperium/h2/issues/463

**Comment by RyanSiu1995** [comment]
Hi there, we encountered similar issue with header too big error as the following.
```
linkerd-proxy [170497.691180s]  WARN ThreadId(01) outbound:accept{peer.addr=10.125.2.107:41600 target.addr=10.125.27.115:8080}: linkerd2_app_core::errors: Failed to proxy request: http2 error: user error: header too big
```
Any idea on how to resolve this?

**Comment by olix0r** [comment]
@RyanSiu1995 It looks like the maximum header (name + value) length is about [16KB](https://github.com/hyperium/h2/blob/5a92f256c0ab63fd12e2180f99e603230955dfbc/src/frame/headers.rs#L17). It appears that this limit is hardcoded and not currently configurable by hyper, though I haven't looked exhaustively. Do you think you're sending individual headers large than 16KB?

If you want to produce diagnostic logs, i'd run the workload with `config.linkerd.io/proxy-log-level: h2=trace,hyper=trace,http=trace,linkerd=info,warn`.

If you have control over these headers, I would examine whether single header values need to be so large. Even if we could increase the limits, large headers are likely to have a negative performance/resource impact.

**Comment by RyanSiu1995** [comment]
Hi @olix0r, yes, we are.
We have the headers larger in 16KB. 
I have asked our internal to change the implementation in the message transfer.
Wonder if it is possible to let us configure the max size of the headers...

**Comment by anorm** [comment]
Seems the header size limitation in h2 is now solved:
https://github.com/hyperium/h2/commit/61b4f8fc34709b7cdfeaf91f3a7a527105c2026b

Hopefully, this will make its way into linkerd2 in the near future :-)

**Comment by olix0r** [comment]
@anorm Thanks for the heads-up! As soon as `h2` is released, we'll get an update from Dependabot. This probably won't make it into `stable-2.11.0`, unfortunately (unless the release happens in the next ~day) -- we're starting to publish release candidates and I'm a little hesitant to change such a fundamental dependency late in the release process. But it will be available on an edge release as soon as `h2` is released :)

**Comment by olix0r** [comment]
We've just upgraded the proxy to use h2 v0.3.6 (https://github.com/linkerd/linkerd2-proxy/pull/1293) which should resolve this issue. Unfortunately, this upgrade did not make it into the stable-2.11.0 release. It should be available in next week's edge-21.10.1 release, though; and we'll try to include this change if/when a stable-2.11.1 update is released.

I'm going to close this for now as there's no work left to do, but please let us know if there are any issues!
